### ISSUE TITLE
Inference on Custom dataset 

### ISSUE THREAD

**Issue by abhigoku10** [tianweiy/CenterPoint#89]
@tianweiy  thanks for sharing the wonderful code base , i wanted to knw which file has to be executed to obtain inference on custom dataset , i have only the point clouds in bin file format which has x,y,z and intensity values. 
THnaks in advance !!

**Comment by andongchen** [comment]
I have change single_inference.py to inference my own dataset like below.
`
import numpy as np
import copy
import json
import os
import sys
import torch
import time 
import pickle
from pyquaternion import Quaternion
from tools.demo_utils import *

from det3d.models import build_detector
from det3d.torchie import Config
from det3d.core.input.voxel_generator import VoxelGenerator

def yaw2quaternion(yaw: float) -> Quaternion:
    return Quaternion(axis=[0,0,1], radians=yaw)

def get_annotations_indices(types, thresh, label_preds, scores):
    indexs = []
    annotation_indices = []
    for i in range(label_preds.shape[0]):
        if label_preds[i] == types:
            indexs.append(i)
    for index in indexs:
        if scores[index] >= thresh:
            annotation_indices.append(index)
    return annotation_indices  


def remove_low_score_nu(image_anno, thresh):
    img_filtered_annotations = {}
    label_preds_ = image_anno["label_preds"].detach().cpu().numpy()
    scores_ = image_anno["scores"].detach().cpu().numpy()
    
    car_indices =                  get_annotations_indices(0, 0.4, label_preds_, scores_)
    truck_indices =                get_annotations_indices(1, 0.4, label_preds_, scores_)
    construction_vehicle_indices = get_annotations_indices(2, 0.4, label_preds_, scores_)
    bus_indices =                  get_annotations_indices(3, 0.3, label_preds_, scores_)
    trailer_indices =              get_annotations_indices(4, 0.4, label_preds_, scores_)
    barrier_indices =              get_annotations_indices(5, 0.4, label_preds_, scores_)
    motorcycle_indices =           get_annotations_indices(6, 0.15, label_preds_, scores_)
    bicycle_indices =              get_annotations_indices(7, 0.15, label_preds_, scores_)
    pedestrain_indices =           get_annotations_indices(8, 0.1, label_preds_, scores_)
    traffic_cone_indices =         get_annotations_indices(9, 0.1, label_preds_, scores_)
    
    for key in image_anno.keys():
        if key == 'metadata':
            continue
        img_filtered_annotations[key] = (
            image_anno[key][car_indices +
                            pedestrain_indices + 
                            bicycle_indices +
                            bus_indices +
                            construction_vehicle_indices +
                            traffic_cone_indices +
                            trailer_indices +
                            barrier_indices +
                            truck_indices
                            ])

    return img_filtered_annotations


class Processor_ROS:
    def __init__(self, config_path, model_path):
        self.points = None
        self.config_path = config_path
        self.model_path = model_path
        self.device = None
        self.net = None
        self.voxel_generator = None
        self.inputs = None
        
    def initialize(self):
        self.read_config()
        
    def read_config(self):
        config_path = self.config_path
        cfg = Config.fromfile(self.config_path)
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.net = build_detector(cfg.model, train_cfg=None, test_cfg=cfg.test_cfg)
        self.net.load_state_dict(torch.load(self.model_path)["state_dict"])
        self.net = self.net.to(self.device).eval()

        self.range = cfg.voxel_generator.range
        self.voxel_size = cfg.voxel_generator.voxel_size
        self.max_points_in_voxel = cfg.voxel_generator.max_points_in_voxel
        self.max_voxel_num = cfg.voxel_generator.max_voxel_num[1]
        self.voxel_generator = VoxelGenerator(
            voxel_size=self.voxel_size,
            point_cloud_range=self.range,
            max_num_points=self.max_points_in_voxel,
            max_voxels=self.max_voxel_num,
        )
        self.lidar_right_to_imu = np.array([[6.1541801362952964e-01, -7.8813947703527232e-01, -9.8403881454022530e-03, 0.48713979,
                                             7.8818583630283823e-01,  6.1543548630651457e-01,  1.4998820818296254e-03, 0.79945409,
                                             4.8740077841227247e-03, -8.6791090114064773e-03,  9.9995045733020604e-01, 0.18072951]]).reshape(3, 4)
        self.lidar_left_to_imu = np.array([[-0.61514498, -0.78795996, 0.02675089, -0.59795754,
                                             0.78798806, -0.61557446, -0.01198092, 0.8311392,
                                             0.02590763, 0.01370934, 0.99957073,   0.19400126]]).reshape(3, 4)

    def run(self, points):
        t_t = time.time()
        print(f"input points shape: {points.shape}")
        num_features = 5        
        self.points = points.reshape([-1, num_features])
        self.points[:, 4] = 0 # timestamp value 
        
        voxels, coords, num_points = self.voxel_generator.generate(self.points)
        num_voxels = np.array([voxels.shape[0]], dtype=np.int64)
        grid_size = self.voxel_generator.grid_size
        coords = np.pad(coords, ((0, 0), (1, 0)), mode='constant', constant_values = 0)
        
        voxels = torch.tensor(voxels, dtype=torch.float32, device=self.device)
        coords = torch.tensor(coords, dtype=torch.int32, device=self.device)
        num_points = torch.tensor(num_points, dtype=torch.int32, device=self.device)
        num_voxels = torch.tensor(num_voxels, dtype=torch.int32, device=self.device)
        
        self.inputs = dict(
            voxels = voxels,
            num_points = num_points,
            num_voxels = num_voxels,
            coordinates = coords,
            shape = [grid_size]
        )
        torch.cuda.synchronize()
        t = time.time()

        with torch.no_grad():
            outputs = self.net(self.inputs, return_loss=False)[0]
    
        # print(f"output: {outputs}")
        
        torch.cuda.synchronize()
        print("  network predict time cost:", time.time() - t)

        outputs = remove_low_score_nu(outputs, 0.45)

        boxes_lidar = outputs["box3d_lidar"].detach().cpu().numpy()
        print("  predict boxes:", boxes_lidar.shape)

        scores = outputs["scores"].detach().cpu().numpy()
        types = outputs["label_preds"].detach().cpu().numpy()

        boxes_lidar[:, -1] = -boxes_lidar[:, -1] - np.pi / 2

        print(f"  total cost time: {time.time() - t_t}")

        return scores, boxes_lidar, types

    def transform(self, points, lidar="left"):
        point_temp = np.concatenate([points[..., :3], np.ones(points.shape[0]).reshape(-1, 1)], axis=1).T
        if lidar == "left":
            point_temp = np.matmul(self.lidar_left_to_imu, point_temp)
        else:
            point_temp = np.matmul(self.lidar_right_to_imu, point_temp)
        points[..., :3] = point_temp.T


if __name__ == "__main__":

    global proc
    ## CenterPoint
    config_path = 'configs/nusc/voxelnet/nusc_centerpoint_voxelnet_01voxel.py'
    model_path = '/data2/chenandong/PycharmProjects/CenterPoint-master/work_dirs/nuse/voxelnet_converted.pth'

    proc_1 = Processor_ROS(config_path, model_path)

    proc_1.initialize()

    root_path_left = "/data2/Datasets/Lidar_camera_synchronization/sweeper/2021/0208/person/velodyne/lidar2/"
    paths_left = os.listdir(root_path_left)
    paths_left.sort()
    root_path_right = "/data2/Datasets/Lidar_camera_synchronization/sweeper/2021/0208/person/velodyne/lidar1/"
    paths_right = os.listdir(root_path_right)
    paths_right.sort()
    index = list(range(len(paths_left)))
    for i, path_left, path_right in zip(index, paths_left, paths_right):

        points_left = np.fromfile(os.path.join(root_path_left, path_left), dtype=np.float32).reshape(-1, 4)
        proc_1.transform(points_left, "left")
        points_right = np.fromfile(os.path.join(root_path_right, path_right), dtype=np.float32).reshape(-1, 4)
        proc_1.transform(points_right, "right")
        points = np.concatenate([points_left, points_right], axis=0)
        points = np.concatenate([points, np.zeros(points.shape[0]).reshape(-1, 1)], axis=1)
        scores, boxes_lidar, types = proc_1.run(points)

        detections = {}
        results = {}
        gt_annos = {}
        detections["scores"] = scores
        detections["box3d_lidar"] = boxes_lidar
        detections["label_preds"] = types
        visual(points[..., :3].T, detections, detections, i)
`

**Comment by andongchen** [comment]
I have change single_inference.py like this to inference my own dataset
`if __name__ == "__main__":

    global proc
    ## CenterPoint
    config_path = 'configs/nusc/voxelnet/nusc_centerpoint_voxelnet_01voxel.py'
    model_path = '/data2/chenandong/PycharmProjects/CenterPoint-master/work_dirs/nuse/voxelnet_converted.pth'

    proc_1 = Processor_ROS(config_path, model_path)
    proc_1.initialize()

    root_path_left = "/data2/Datasets/Lidar_camera_synchronization/sweeper/2021/0208/person/velodyne/lidar2/"
    paths_left = os.listdir(root_path_left)
    paths_left.sort()
    for i, path_left in enumerate(paths_left):
        points_left = np.fromfile(os.path.join(root_path_left, path_left), dtype=np.float32).reshape(-1, 4)
        points = np.concatenate([points_left, np.zeros(points_left.shape[0]).reshape(-1, 1)], axis=1)
        scores, boxes_lidar, types = proc_1.run(points)

        detections = {}
        results = {}
        gt_annos = {}
        detections["scores"] = scores
        detections["box3d_lidar"] = boxes_lidar
        detections["label_preds"] = types
        visual(points[..., :3].T, detections, detections, i)
   `

**Comment by abhigoku10** [comment]
@tianweiy can i use the default nusc config file for custom data or should i use "nusc_centerpoint_pp_02voxel_circle_nms_demo.py"  which you have mentioned in single_inference.py

**Comment by abhigoku10** [comment]
Thanks !!

**Comment by tianweiy** [comment]
I will add a script next week.  I think we want some script that basically turns a NumPy point cloud into a few numpy boxes. In this way, we can also do further optimization (fp16, tensorrt, bn/conv fusion, etc..).

**Comment by tianweiy** [comment]
I will fix this today. It should not be moved to cpu

**Comment by tianweiy** [comment]
I think it should work now?

**Comment by YoushaaMurhij** [comment]
> I will add a script next week. I think we want some script that basically turns a NumPy point cloud into a few numpy boxes. In this way, we can also do further optimization (fp16, tensorrt, bn/conv fusion, etc..).

Yes the _voxel_generater_ is the bottleneck for inference on light devices like (Jetson Xavier.. etc). I tried to reduce the input point cloud size. it does improve the inference speed but it is not always suitable. If this process is moved to GPU,  there will be a noticeable improvement.

**Comment by tianweiy** [comment]
yes, the thing you need to do is just replace the current voxelization https://github.com/tianweiy/CenterPoint/blob/8ddae3082151aafdfaa0f5a7b852eb13c10c031b/tools/simple_inference_waymo.py#L58 with a faster one. In theory, it should only takes 2-4 ms for whatever large point clouds.

**Comment by tianweiy** [comment]
actually you can do some trick like this  for the voxelization on GPU 

```
def voxelization(points, pc_range, voxel_size):    
    keep = (points[:, 0] >= pc_range[0]) * (points[:, 0] <= pc_range[3]) * \
        (points[:, 1] >= pc_range[1]) * (points[:, 1] <= pc_range[4]) * \
            (points[:, 2] >= pc_range[2]) * (points[:, 2] <= pc_range[5])
    points = points[keep, :]    
    coords = ((points[:, [2, 1, 0]] - pc_range[[2, 1, 0]]) /  voxel_size[[2, 1, 0]]).to(torch.int64)
    unique_coords, inverse_indices = coords.unique(return_inverse=True, dim=0)

    voxels = scatter_mean(points, inverse_indices, dim=0)
    return voxels, unique_coords
```
assuming points is already on GPU. scatter_mean is from torch_scatter https://github.com/rusty1s/pytorch_scatter and doesn't work with pytorch 1.1 to my knowledge. 

 then directly pass the output into the VoxelNet backbone by replacing 

https://github.com/tianweiy/CenterPoint-dev/blob/pre_detr/det3d/models/detectors/voxelnet.py

with 

```
    def extract_feat(self, data):
        input_features = data["features"]
        x, voxel_feature = self.backbone(
            input_features, data["coors"], data["batch_size"], data["input_shape"]
        )
        if self.with_neck:
            x = self.neck(x)

        return x, voxel_feature
```

**Comment by tianweiy** [comment]
the above step takes 3 ms for super large point cloud. GPU Voxelization (dynamic voxelization + dynamic scatter)For PointPillars is much harder and I don't think there are simple tricks for this.  I personally just use VoxelNet as it is not much slower with good accuracy.

**Comment by YoushaaMurhij** [comment]
@tianweiy, when calling [voxel_generator](https://github.com/tianweiy/CenterPoint/blob/8ddae3082151aafdfaa0f5a7b852eb13c10c031b/tools/simple_inference_waymo.py#L58) for the first time, it take very long time after that it comes back to normal. Can we get rid of this first iteration long delay?

**Comment by tianweiy** [comment]
I haven't met that but if it is really just the first iteration. You can try run this first iteration in the initialization ? And then it could work well?

**Comment by YoushaaMurhij** [comment]
> I haven't met that but if it is really just the first iteration. You can try run this first iteration in the initialization ? And then it could work well?

I will try that. these are the times.
```
  network voxel_gen time cost: 0.7885196208953857
  network inference time cost: 0.03793001174926758
  total callback time:         0.8492271900177002
  network voxel_gen time cost: 0.001783609390258789
  network inference time cost: 0.025906085968017578
  total callback time:         0.03858590126037598
  network voxel_gen time cost: 0.0018002986907958984
  network inference time cost: 0.025873184204101562
  total callback time:         0.03880882263183594
```

**Comment by tianweiy** [comment]
which file are you using? The ros script? The ros script uses a python voxel generator which needs time to compile at the beginning to my knowledge. You can replace with the voxel generator in spconv like I did in simple_waymo_inference

**Comment by YoushaaMurhij** [comment]
Thanks for mentioning that!
I got this Assertion Error when replacing the voxel generator:
```python
  voxels, coords, num_points = self.voxel_generator.generate(self.points)
  File "/usr/local/lib/python3.8/dist-packages/spconv/utils/__init__.py", line 170, in generate
    res = points_to_voxel(points, self._voxel_size,
  File "/usr/local/lib/python3.8/dist-packages/spconv/utils/__init__.py", line 69, in points_to_voxel
    assert block_filtering is False
AssertionError
```

**Comment by tianweiy** [comment]
I saw this in the past. It is some version problem. It should be easily fixable. Please take a look at simple_waymo_inference and check spconv code at this line. I can take a look tonight if you still have a problem with this.

**Comment by tianweiy** [comment]
@YoushaaMurhij  Have you solved this? I think it is related to spconv version like there are two voxelizer in the spconv master branch, voxelizerv1, voxelizerv2, they have a little different APIs and then you may face this assert issue with one specific version if some arguments are not set correctly. You can also use the default voxelizer in this repo, but in my experiment, this cpu voxelizer in numpy is slower than the c++ one in spconv

**Comment by YoushaaMurhij** [comment]
No. I am facing a problem with the dependencies. This [spconv commit](https://github.com/traveller59/spconv/commit/73427720a539caf9a44ec58abe3af7aa9ddb8e39) seems only to run with cuda10.
Have you tried the latest spconv commit?

**Comment by tianweiy** [comment]
yes, latest one should work (I tried this). Maybe use the voxelizerv2 ?

**Comment by tianweiy** [comment]
change this line https://github.com/tianweiy/CenterPoint/blob/9fdf572e854d20d58f4b0b8ef9f5785a5d39147a/tools/simple_inference_waymo.py#L58 to 
```
        voxel_output = voxel_generator.generate(points)
        if isinstance(voxel_output, dict):
            voxels, coordinates, num_points = \
                voxel_output['voxels'], voxel_output['coordinates'], voxel_output['num_points_per_voxel']
        else:
            voxels, coordinates, num_points = voxel_output
```

**Comment by YoushaaMurhij** [comment]
Thanks. I tried it and it works!
But it seems to work a little slower than the cpu voxelizer which is strange!

**Comment by YoushaaMurhij** [comment]
@tianweiy , It is not clear how the voxelizer chooses the points to include them in a certain voxel/pillar. Is it randomly adding the first max_points that are in this voxel/pillar?
for pointpillars situation it will be a problem as the pillars height is 6 or 8 meters. for pedestrian detection the pillar will contains only some upper or lower points.  Can you correct me please? Can you refer to the code that does this task? 
Thank you!

**Comment by tianweiy** [comment]
random, you can increase max_points

**Comment by tianweiy** [comment]
https://github.com/tianweiy/CenterPoint/blob/9fdf572e854d20d58f4b0b8ef9f5785a5d39147a/det3d/ops/point_cloud/point_cloud_ops.py#L112

**Comment by YoushaaMurhij** [comment]
![Screenshot from 2021-04-20 17-35-06](https://user-images.githubusercontent.com/16229594/115705678-80a91d80-a375-11eb-99e2-0805bbddb9ee.png)
![Screenshot from 2021-04-21 12-36-12](https://user-images.githubusercontent.com/16229594/115705691-830b7780-a375-11eb-91d9-0b4063f620e1.png)
In the first image the voxel takes the first n_max points and in the second one I shuffled the points. (The white points are the ones that are included in voxels while the colorful ones are the original pointcloud)
when the pedestrian is very close to the lidar like in these images, the network is unable to detect it. Increasing the max_points in every voxel/pillar doesn't solve that.

**Comment by muzi2045** [comment]
maybe you need the dynamic voxel to process this problem.

**Comment by YoushaaMurhij** [comment]
> maybe you need the dynamic voxel to process this problem.

@muzi2045 Is this "dynamic voxel" implemented somewhere here?

**Comment by YoushaaMurhij** [comment]
@tianweiy, I saw that your Waymo-PointPillars model (1st stage) achieved 19 FPS. Can you tell on which device and the size of the input pointcloud? I tried it on Waymo-pointclouds and got less than 10 FPS!

Thanks!
